samples/mt_gc_test.cr

```diff
@@ -0,0 +1,229 @@
+# The experiment manually allocates multiple times a queue of fibers
+# on a list of worker threads. The main thread does the orchestration.
+# The fibers will grow and shrink their stack size by doing recursion.
+# In each level of the stack an amount of dummy objects are allocated.
+# These objects will be released by the GC eventually.
+
+require "option_parser"
+require "benchmark"
+
+lib LibC
+  fun fflush(b : Void*)
+end
+
+class Foo
+  @@collections = Atomic(Int32).new(0)
+
+  @data = StaticArray(Int32, 8).new(0)
+
+  def finalize
+    @@collections.add(1)
+  end
+
+  def self.collections
+    @@collections.get
+  end
+end
+
+class Context
+  enum State
+    Wait
+    Run
+  end
+
+  property expected_depth : Int32 = 0
+  @worker_fibers = Array(Fiber).new(0)
+
+  def initialize(@fibers : Int32, @threads : Int32, @log : Bool)
+    @fibers_reached = Atomic(Int32).new(0)
+    @threads_reached = Atomic(Int32).new(0)
+    @fiber_depths = Array(Int32).new(@fibers, 0)
+    @pending_fibers_queues = Array(Array(Fiber)).new(@threads) { Array(Fiber).new }
+    @threads_state = Atomic(State).new(State::Wait)
+
+    # Create worker fibers but do not start them yet.
+    # Each fiber will try to reach the `expected_depth` value
+    # by increasing or decreasing its callstack.
+    @fibers.times do |index|
+      @worker_fibers << Fiber.new("f:#{index}") do
+        Context.fiber_run(self, index, 1)
+      end
+    end
+
+    # Create worker threads.
+    # they will perform operations when `@threads_state == :run`
+    # otherwise they will remain in a tight busy loop.
+    # See Context#create_thread
+    @threads.times { |index| create_thread(index) }
+  end
+
+  def self.fiber_run(context, fiber_index, depth)
+    context.set_fiber_depth(fiber_index, depth)
+
+    # allocate a bunch of objects in the stack
+    # some should be released fast
+    10.times do
+      foo = Foo.new
+    end
+    foo = Foo.new
+
+    # increase/decrease stack depending on the expected_depth
+    # when reached, notify and yield control
+    while true
+      if context.expected_depth < depth
+        return
+      elsif context.expected_depth > depth
+        fiber_run(context, fiber_index, depth + 1)
+      else
+        context.notifify_depth_reached
+        context.yield
+      end
+    end
+  end
+
+  def set_fiber_depth(index, depth)
+    @fiber_depths[index] = depth
+  end
+
+  def run_until_depth(phase, depth)
+    # make all fibers reach a specific depth
+    log "#{phase}: expected_depth: #{depth}"
+
+    @expected_depth = depth
+    @fibers_reached.set(0)
+    @threads_reached.set(0)
+
+    # allocate fibers on each thread queue.
+    @pending_fibers_queues.each &.clear
+    @worker_fibers.dup.shuffle!.each_with_index do |f, index|
+      @pending_fibers_queues[index % @threads] << f
+    end
+
+    @threads_state.set(State::Run)
+
+    # spin wait for all fibers to finish
+    while (c = @fibers_reached.get) < @fibers
```

Comment:
What `c = ...` assignment is for? ditto below.

Reply:
old puts debugging 🙈